### PR TITLE
DataSourceGeneric resource

### PR DESCRIPTION
**Pull request by jwierzbo**
There are plenty third party Datasources plugins for Grafana.
To manage them by REST API we shouldn't stick to the fixed list of fields in Datasource class.

To solve above case, I've prepared DataSourceGeneric class which allow to manage any type of Grafana Datasource.
I've use this also in my custom Grafana terraform provider - https://github.com/jwierzbo/terraform-provider-grafanads
To solve following issue: https://github.com/grafana/terraform-provider-grafana/issues/63

### CODE REVIEW

**Comment by aknuds1** [comment]
What do you think @marefr?

File: datasource_generic.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+	OrgId     int64 `json:"orgId,omitempty"`
+	IsDefault bool  `json:"isDefault"`
+
+	BasicAuth     bool   `json:"basicAuth"`
+	BasicAuthUser string `json:"basicAuthUser,omitempty"`
+	// Deprecated in favor of secureJsonData.basicAuthPassword
+	BasicAuthPassword string `json:"basicAuthPassword,omitempty"`
+
+	JSONData       JsonData `json:"jsonData,omitempty"`
+	SecureJSONData JsonData `json:"secureJsonData,omitempty"`
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
	ID     int64  `json:"id,omitempty"`
	Name   string `json:"name"`
	Type   string `json:"type"`
	URL    string `json:"url"`
	Access string `json:"access"`

	Database string `json:"database,omitempty"`
	User     string `json:"user,omitempty"`
	// Deprecated in favor of secureJsonData.password
	Password string `json:"password,omitempty"`

	OrgID     int64 `json:"orgId,omitempty"`
	IsDefault bool  `json:"isDefault"`

	BasicAuth     bool   `json:"basicAuth"`
	BasicAuthUser string `json:"basicAuthUser,omitempty"`
	// Deprecated in favor of secureJsonData.basicAuthPassword
	BasicAuthPassword string `json:"basicAuthPassword,omitempty"`

	JSONData       JsonData `json:"jsonData,omitempty"`
	SecureJSONData JsonData `json:"secureJsonData,omitempty"`
```

File: datasource_generic.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+type JsonData map[string]interface{}
```
Comment (aknuds1) [inline datasource_generic.go]: Does this type have to be exported?
```suggestion
type JSONData map[string]interface{}
```

File: datasource_generic.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return 0, err
+	}
+	if resp.StatusCode != 200 {
+		return 0, errors.New(resp.Status)
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
		return 0, fmt.Errorf("creating data source failed: %s", resp.Status)
```

File: datasource_generic.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+	req, err := c.newRequest("POST", "/api/datasources", nil, bytes.NewBuffer(data))
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return 0, err
+	}
+	if resp.StatusCode != 200 {
```
Comment (aknuds1) [inline datasource_generic.go]: I think all 2xx status codes should be OK:
```suggestion
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
```

File: datasource_generic.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+		return 0, errors.New(resp.Status)
+	}
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
	}
	defer resp.Body.Close()
```

File: datasource_generic.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+	data, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return 0, err
+	}
+
+	result := struct {
+		Id int64 `json:"id"`
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
		ID int64 `json:"id"`
```

File: datasource_generic.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+	req, err := c.newRequest("PUT", path, nil, bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
```

File: datasource_generic.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+		return errors.New(resp.Status)
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
		return fmt.Errorf("updating data source failed: %s", resp.Status)
```

File: datasource_generic.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	req, err := c.newRequest("PUT", path, nil, bytes.NewBuffer(data))
+	if err != nil {
+		return err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return err
+	}
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
	}
	defer resp.Body.Close()
```

File: datasource_generic.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+		return nil, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != 200 {
+		return nil, errors.New(resp.Status)
+	}
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("getting data source failed: %s", resp.Status)
	}
```

File: datasource_generic.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+		return err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != 200 {
+		return errors.New(resp.Status)
+	}
```
Comment (aknuds1) [inline datasource_generic.go]:
```suggestion
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("failed deleting data source: %s", resp.Status)
	}
```

File: datasource_generic_test.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+		},
+		SecureJSONData: JsonData{
+			"AccessKey": "123",
+			"SecretKey": "456",
+		},
+	}
+
+	created, err := client.NewDataSourceGeneric(ds)
+	if err != nil {
+		t.Error(err)
```
Comment (aknuds1) [inline datasource_generic_test.go]:
```suggestion
		t.Fatal(err)
```

File: datasource_generic_test.go [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+		JSONData: JsonData{
+			"HttpMethod":   "POST",
+			"QueryTimeout": "60s",
+			"TimeInterval": "1m",
+		},
+	}
+
+	created, err := client.NewDataSourceGeneric(ds)
+	if err != nil {
+		t.Error(err)
```
Comment (aknuds1) [inline datasource_generic_test.go]:
```suggestion
		t.Fatal(err)
```

File: datasource_generic_test.go [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	t.Log(pretty.PrettyFormat(created))
+
+	if created != 1 {
+		t.Error("datasource creation response should return the created datasource ID")
```
Comment (aknuds1) [inline datasource_generic_test.go]:
```suggestion
		t.Error("data source creation response should return the created data source ID")
```

**Review by aknuds1 — DISMISSED** [review]
I don't know yet if this is a desired feature, but left suggestions for issues I saw.

**Comment by marefr** [comment]
It make sense, but not optimal to have two different ways to operate on data sources. However, doesn't seems like there's any releases made so could be tricky to introduce breaking changes and remove the non-generic.

**Comment by jwierzbo** [comment]
Should I merge with current master and apply @aknuds1 changes or you want me to close this PR?

File: datasource_generic.go (hunk 1 above)
Comment (jwierzbo) [reply datasource_generic.go]: This is copy/paste of https://github.com/jwierzbo/grafana-api-golang-client/blob/4cd08a254b133d2ecabaa6c052008646f6b25524/datasource.go#L11

Let's stay consistent

File: datasource_generic.go (hunk 2 above)
Comment (jwierzbo) [reply datasource_generic.go]: This is copy/paste of https://github.com/jwierzbo/grafana-api-golang-client/blob/4cd08a254b133d2ecabaa6c052008646f6b25524/datasource.go#L30

Let's stay consistent

File: datasource_generic.go (hunk 6 above)
Comment (jwierzbo) [reply datasource_generic.go]: Let's stay consistent - check my previous comments

**Comment by jwierzbo** [comment]
@aknuds1 could you review again?

**Comment by tiwood** [comment]
@aknuds1, anything we can do to help?
We would love use the official Grafana Terraform provider, but are still blocked by this dependency.

**Comment by jwierzbo** [comment]
@aknuds1 should I close this PR?

**Comment by aymericDD** [comment]
Any news about this PR? It could be very useful :+1: 
Thanks for your work @jwierzbo :heart:

**Comment by julienduchesne** [comment]
As said here: https://github.com/grafana/terraform-provider-grafana/issues/197#issuecomment-1047012476. This will be supported through a new client

**Comment by julienduchesne** [comment]
Supported as part of https://github.com/grafana/grafana-api-golang-client/pull/107

**Pull request closed without merging**